FeatureEngine.py

```diff
@@ -42,37 +44,243 @@ def map_labels_to_invoice_OCR(self, data_path: str, verbose: bool = False):
+            inner_rect_width = max(
+                0,
+                outer_rect_width - (t1.coordinates["width"] + t2.coordinates["width"]),
+            )
+            inner_rect_height = max(
+                0,
+                outer_rect_heigth
+                - (t1.coordinates["height"] + t2.coordinates["height"]),
+            )
+            pixel_dist = math.sqrt(inner_rect_width ** 2 + inner_rect_height ** 2)
+            return pixel_dist
+
+        # checks if two tokens are aligned vertically within a margin of error (checks midpoint, left boundary, right boundary)
+        def is_vert_aligned(t1, t2, moe):
+            if abs(t1.coordinates["x"] - t2.coordinates["x"]) < moe:
+                return True
+            if (
+                abs(
+                    (t1.coordinates["x"] + t1.coordinates["width"])
+                    - (t2.coordinates["x"] + t2.coordinates["width"])
+                )
+                < moe
+            ):
+                return True
+            t1_midpt_x = t1.coordinates["x"] + (t1.coordinates["width"] / 2)
+            t2_midpt_x = t2.coordinates["x"] + (t2.coordinates["width"] / 2)
+            if abs(t1_midpt_x - t2_midpt_x) < moe:
+                return True
+            return False
+
+        # checks if two tokens are aligned vertically within a margin of error (checks midpoint, top boundary, bottom boundary)
+        def is_hori_aligned(t1, t2, moe):
+            if abs(t1.coordinates["y"] - t2.coordinates["y"]) < moe:
+                return True
+            if (
+                abs(
+                    (t1.coordinates["y"] + t1.coordinates["height"])
+                    - (t2.coordinates["y"] + t2.coordinates["height"])
+                )
+                < moe
+            ):
+                return True
+            t1_midpt_y = t1.coordinates["y"] + (t1.coordinates["height"] / 2)
+            t2_midpt_y = t2.coordinates["y"] + (t2.coordinates["height"] / 2)
+            if abs(t1_midpt_y - t2_midpt_y) < moe:
+                return True
+            return False
+
+        #############################################################################################################################
+
         features = {}
 
-        ### Datatype-related features ###
-        # TODO: Implement regexes to determine datatype of this token - DATE, MONEY, NUMBER
+        # number of characters and words in token text string
+        features["char_count"] = len(token.text)
+        features["word_count"] = len(token.text.split(" "))
 
-        ### Vertical location feature ###
-        # TODO: Calculate distance from the top of the image
+        # height and width of token
+        features["height"] = token.coordinates["height"]
+        features["width"] = token.coordinates["width"]
 
-        ### Date-alignment related features ###
-        date_related_tokens = invoicePage.search_tokens("date")
+        # distance to edges of page (to nearest point on the box)
```

Comment:
Wonder if it is more meaningful to calculate distance to the instance of a token because invoices are spaced out differently. So rather than looking at distance from the edges, you look at where the invoice actually starts

Reply:
i dont understand what you mean here

Reply:
solved